### PR TITLE
feat(xds): Add Kuma service discovery to the ScrapeConfig CRD

### PR DESCRIPTION
**Pull request by dongjiang1989**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

support kuma xds
ref: https://github.com/prometheus/prometheus/blob/main/docs/configuration/configuration.md#kuma_sd_config

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Add Kuma service discovery to the ScrapeConfig CRD
```

### CODE REVIEW

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 1]
```diff
@@ -675,3 +678,53 @@ type DigitalOceanSDConfig struct {
...
 }
+
+// KumaSDConfig allow retrieving scrape targets from Kuma's control plane.
+// See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#kuma_sd_config
+// +k8s:openapi-gen=true
+type KumaSDConfig struct {
+	// Address of the Kuma Control Plane's MADS xDS server.
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Server *string `json:"server"`
```
Comment (ArthurSens) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:689]:
```suggestion
	Server string `json:"server"`
```

Let's not use pointers for required fields.

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

**Review by ArthurSens — COMMENTED** [review]
One small nit.

Also, just wanted to let you and @slashpai know that we'll have conflicts between this PR and https://github.com/prometheus-operator/prometheus-operator/pull/6333. Let's coordinate the merges if we get both ready at the same time :)

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 1 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: Thanks. Got it. change `*string` to `string`

**Comment by dongjiang1989** [comment]
> One small nit.
> 
> Also, just wanted to let you and @slashpai know that we'll have conflicts between this PR and #6333. Let's coordinate the merges if we get both ready at the same time :)

Got it. thanks @ArthurSens

**Comment by dongjiang1989** [comment]
@slashpai  PTAL, thanks

**Comment by slashpai** [comment]
Can you rebase since https://github.com/prometheus-operator/prometheus-operator/pull/6401 is merged

I see you have one more PR so you may need to rebase again based on which ever gets merged first

**Comment by slashpai** [comment]
Also if possible squash the commits, there is option for squashing with merge but still it look long commit message then with 8 commit messages :) probably keep it short like `feat: add support for Kuma SD in ScrapeConfig`

**Comment by slashpai** [comment]
@dongjiang1989 I have merged https://github.com/prometheus-operator/prometheus-operator/pull/6408 so this would need rebase

Since [scrape config is part of GSoc](https://github.com/prometheus-operator/community/blob/main/mentoring/gsoc/2024/project_ideas.md#extending-scrapeconfig-surface) there are many PRs related to that as well so based on whichever gets merged first others will get merge conflict. Sometimes can end up creating duplicate PRs. I would suggest to create a github issue first or mention in #prometheus-operator-dev channel in Kubernetes slack incase if you are working on any Scrape Config SD.

Thank you very much for working on this and having patience :)

**Comment by slashpai** [comment]
@dongjiang1989 we have a release tomorrow so incase if you get time before that can you rebase so we can get this in?

**Comment by dongjiang1989** [comment]
> @dongjiang1989 we have a release tomorrow so incase if you get time before that can you rebase so we can get this in?

Thanks. I got it.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 2]
```diff
@@ -675,3 +678,53 @@ type DigitalOceanSDConfig struct {
...
+	Server string `json:"server"`
+
+	// Client id is used by Kuma Control Plane to compute Monitoring Assignment for specific Prometheus backend.
+	// +optional
+	ClientID *string `json:"clientID,omitempty"`
+
+	// The time to wait between polling update requests.
+	// +optional
+	RefreshInterval *v1.Duration `json:"refreshInterval,omitempty"`
+
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: Thanks. Done

File: pkg/prometheus/resource_selector.go [hunk 3]
```diff
@@ -1028,3 +1033,33 @@ func (rs *ResourceSelector) validateDigitalOceanSDConfigs(ctx context.Context, s
...
+		if err := rs.store.AddBasicAuth(ctx, sc.GetNamespace(), config.BasicAuth, configKey); err != nil {
+			return fmt.Errorf("[%d]: %w", i, err)
+		}
+
+		if config.ProxyConfig != nil {
+			if err := validateProxyConfig(ctx, config.ProxyConfig, rs.store, sc.GetNamespace()); err != nil {
+				return fmt.Errorf("[%d]: %w", i, err)
+			}
+		}
+
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/prometheus/resource_selector.go]: Thanks. Done

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 4]
```diff
@@ -675,3 +678,53 @@ type DigitalOceanSDConfig struct {
...
+type KumaSDConfig struct {
+	// Address of the Kuma Control Plane's MADS xDS server.
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Server string `json:"server"`
+
+	// Client id is used by Kuma Control Plane to compute Monitoring Assignment for specific Prometheus backend.
+	// +optional
+	ClientID *string `json:"clientID,omitempty"`
+
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: Thanks. Done

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 5]
```diff
@@ -675,3 +678,53 @@ type DigitalOceanSDConfig struct {
...
+// KumaSDConfig allow retrieving scrape targets from Kuma's control plane.
+// See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#kuma_sd_config
+// +k8s:openapi-gen=true
+type KumaSDConfig struct {
+	// Address of the Kuma Control Plane's MADS xDS server.
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Server string `json:"server"`
+
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: Thanks. Done

File: pkg/prometheus/resource_selector_test.go [hunk 6]
```diff
@@ -2172,6 +2206,26 @@ func TestSelectScrapeConfigs(t *testing.T) {
 			},
 			selected: true,
 		},
+		{
+			scenario: "Kuma SD config with invalid TLS config with invalid CA data",
```
Comment (slashpai) [inline pkg/prometheus/resource_selector_test.go:2210]: nit: lets keep kuma sd tests together

File: pkg/prometheus/resource_selector_test.go [hunk 7]
```diff
@@ -2192,6 +2246,29 @@ func TestSelectScrapeConfigs(t *testing.T) {
 			},
 			selected: false,
 		},
+		{
+			scenario: "Kuma SD config with valid proxy settings",
```
Comment (slashpai) [inline pkg/prometheus/resource_selector_test.go:2250]: nit: keep the kuma sd tests together

**Comment by dongjiang1989** [comment]
@slashpai  all nits fixed.

File: pkg/prometheus/resource_selector_test.go (hunk 6 above)
Comment (dongjiang1989) [reply pkg/prometheus/resource_selector_test.go]: Done

File: pkg/prometheus/resource_selector_test.go (hunk 7 above)
Comment (dongjiang1989) [reply pkg/prometheus/resource_selector_test.go]: Done

File: pkg/prometheus/promcfg.go [hunk 8]
```diff
@@ -3320,6 +3320,72 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+	if len(sc.Spec.KumaSDConfigs) > 0 {
+		configs := make([][]yaml.MapItem, len(sc.Spec.KumaSDConfigs))
+		for i, config := range sc.Spec.KumaSDConfigs {
+			assetStoreKey := fmt.Sprintf("scrapeconfig/%s/%s/kumasdconfig/%d", sc.GetNamespace(), sc.GetName(), i)
+			configs[i] = cg.addBasicAuthToYaml(configs[i], assetStoreKey, store, config.BasicAuth)
+			configs[i] = cg.addSafeAuthorizationToYaml(configs[i], fmt.Sprintf("scrapeconfig/auth/%s/%s/kumasdconfig/%d", sc.GetNamespace(), sc.GetName(), i), store, config.Authorization)
+			configs[i] = cg.addOAuth2ToYaml(configs[i], config.OAuth2, store.OAuth2Assets, assetStoreKey)
+			configs[i] = cg.addProxyConfigtoYaml(ctx, configs[i], sc.GetNamespace(), store, config.ProxyConfig)
+
+			if config.Server != "" {
```
Comment (slashpai) [inline pkg/prometheus/promcfg.go:3333]: `Server` is a required field right, we don't need to check for empty here

File: pkg/prometheus/promcfg.go [hunk 9]
```diff
@@ -3320,6 +3320,72 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+					Key:   "enable_http2",
+					Value: config.EnableHTTP2,
+				})
+			}
+
+			if config.TLSConfig != nil {
+				configs[i] = addSafeTLStoYaml(configs[i], sc.GetNamespace(), *config.TLSConfig)
+			}
+
+			if config.TLSConfig != nil {
```
Comment (slashpai) [inline pkg/prometheus/promcfg.go:3358]: I see this is repeated check here

File: pkg/prometheus/promcfg.go [hunk 10]
```diff
@@ -3320,6 +3320,72 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 		})
 	}
 
+	//KumaSDConfig
```
Comment (slashpai) [inline pkg/prometheus/promcfg.go:3323]:
```suggestion
	// KumaSDConfig
```

**Review by slashpai — COMMENTED** [review]
Sorry something I missed in preview review

File: pkg/prometheus/promcfg.go (hunk 9 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go]: Thanks. Fixed

File: pkg/prometheus/promcfg.go (hunk 10 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go]: Thanks. Fixed

File: pkg/prometheus/promcfg.go (hunk 8 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go]: Done

**Review by slashpai — APPROVED** [review]

**Comment by dongjiang1989** [comment]
@slashpai  Please recheck. thanks

**Comment by slashpai** [comment]
For Server we can add a validation as well but this can be a followup PR, it needs to add in resource_selector.go

see https://github.com/prometheus/prometheus/blob/main/discovery/xds/kuma.go#L78-L85

File: pkg/prometheus/resource_selector.go [hunk 11]
```diff
@@ -1025,6 +1030,33 @@ func (rs *ResourceSelector) validateDigitalOceanSDConfigs(ctx context.Context, s
 	return nil
 }
 
+func (rs *ResourceSelector) validateKumaSDConfigs(ctx context.Context, sc *monitoringv1alpha1.ScrapeConfig) error {
+	for i, config := range sc.Spec.KumaSDConfigs {
```
[the remark this answers is not in the document]
Comment (slashpai) [reply pkg/prometheus/resource_selector.go:1034]: cc: @dongjiang1989

[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/prometheus/resource_selector.go:1034]: Thanks. Will add another pr to solve it.

**Pull request merged**